feat: Add Terraform for automated infrastructure deployment

a-custodio opened the pull request:
This PR introduces a complete Terraform configuration to automate the deployment of the Zombie Projects Watcher. It manages all required GCP resources, including the Cloud Function, GCS buckets, Secret Manager secrets, and IAM permissions, enabling a fully version-controlled and repeatable setup.

a-custodio committed:
feat: Add Terraform for automated end-to-end deployment

Introduces a complete Terraform configuration to automate the deployment of all required GCP resources. This includes GCS buckets, Secret Manager secret, IAM permissions, the Cloud Function, and the Cloud Scheduler job.

a-custodio committed:
Adding .tfvars files in gitignore

a-custodio committed:
refactor(terraform): Update function resource for local config

Modifies the function's deployment process to bundle the config.yaml file directly, removing the dependency on GCS environment variables for now.

a-custodio committed:
docs: Revise documentation for new Terraform architecture

- Replaces manual gcloud commands with a complete Terraform deployment guide.
- Updates requirements, APIs, and IAM roles.
- Adds new instructions for local development.

a-custodio committed:
Adding a new blank line in the end of the code

guilherme-hosoda-cit commented on README.md:
```diff
@@ -50,71 +47,28 @@ gcloud services enable \
     secretmanager.googleapis.com \
-    --project ${PROJECT_ID}
+    storage.googleapis.com \
+    --project <YOUR-PROJECT-ID>
```
NIT: The old structure is easier to use since the variable is set in the beginning of the code block.

guilherme-hosoda-cit commented on terraform/.terraform.lock.hcl:
Add this file to the `.gitignore`

guilherme-hosoda-cit commented on terraform/function.tf:
NIT: Rename the folder to `infra` instead of `terraform`

guilherme-hosoda-cit commented on terraform/function.tf:
```diff
@@ -0,0 +1,69 @@
+data "archive_file" "source" {
+  type        = "zip"
+  source_dir  = "../"
```
Use `${path.module}` variable to avoid an invalid file when running in a different cwd.

```suggestion
  source_dir  = "${path.module}/../"
```

<https://developer.hashicorp.com/terraform/language/expressions/references#path-module>

guilherme-hosoda-cit commented on terraform/function.tf:
Run `terraform fmt --recursive` to format all files.

guilherme-hosoda-cit commented on terraform/function.tf:
```diff
@@ -0,0 +1,69 @@
+      service_account_email = var.service_account_email
+    }
+  }
+}
```
Insert new line in the end

guilherme-hosoda-cit commented on terraform/main.tf:
```diff
@@ -0,0 +1,18 @@
+  }
+}
+
+provider "google" {
```
Move the `provider` block to the `providers.tf` file.

guilherme-hosoda-cit commented on terraform/main.tf:
```diff
@@ -0,0 +1,18 @@
+    }
+  }
+
+  backend "gcs" {
```
Move the `backend` block to the `backend.tf` file.

guilherme-hosoda-cit commented on terraform/main.tf:
```diff
@@ -0,0 +1,18 @@
+terraform {
```
Move the `terraform` block to the `versions.tf` file.

guilherme-hosoda-cit commented on terraform/iam.tf:
```diff
@@ -0,0 +1,23 @@
+resource "google_project_iam_member" "secret_accessor" {
+  project = var.project_id
+  role    = "roles/secretmanager.secretAccessor"
+  member  = "serviceAccount:${var.service_account_email}"
```
Why the service account is a variable? Wouldn't it be better to create the service account in the code?

guilherme-hosoda-cit commented on terraform/main.tf:
```diff
@@ -0,0 +1,18 @@
+provider "google" {
+  project = var.project_id
+  region  = var.region
+}
```
Missing blank line

guilherme-hosoda-cit commented on terraform/outputs.tf:
```diff
@@ -0,0 +1,4 @@
+output "function_url" {
+  description = "The invocation URL of the deployed Cloud Function."
+  value       = google_cloudfunctions2_function.zombie_watcher_function.service_config[0].uri
+}
```
Missing blank line.

guilherme-hosoda-cit commented on terraform/secret_manager.tf:
```diff
@@ -0,0 +1,12 @@
+resource "google_secret_manager_secret_version" "webhook_secret_version" {
+  secret      = google_secret_manager_secret.webhook_secret.id
+  secret_data = var.chat_webhook_url
+}
```
Missing blank line

guilherme-hosoda-cit commented on terraform/gcs.tf:
```diff
@@ -0,0 +1,16 @@
+resource "google_storage_bucket" "config_bucket" {
+  name          = var.config_bucket_name
```
NIT: it would be better to create the bucket name dynamically, using a random_id as suffix.

```hcl
resource "random_id" "config_bucket_suffix" {
  byte_length = 2
}

resource "google_storage_bucket" "config_bucket" {
  name = "zombie-watcher-config-${var.project_id}-${random_id.config_bucket_suffix.hex}"
 ...
}
```

guilherme-hosoda-cit commented on terraform/gcs.tf:
```diff
@@ -0,0 +1,16 @@
+resource "google_storage_bucket_object" "config_file" {
+  name   = "config.yaml"
+  bucket = google_storage_bucket.config_bucket.name
+  source = "../config.yaml"
```
Use `${path.module}`

```suggestion
  source = "${path.module}/../config.yaml"
```

guilherme-hosoda-cit commented on terraform/variables.tf:
```diff
@@ -0,0 +1,32 @@
+variable "config_bucket_name" {
+  description = "The name for the GCS bucket that will store the config.yaml file."
+  type        = string
+}
```
Missing blank line.

a-custodio committed:
docs: Improve usability of gcloud services command

a-custodio committed:
chore: Ignore Terraform lock file

Adds .terraform.lock.hcl to the .gitignore file.

a-custodio committed:
chore: Rename terraform directory to infra

a-custodio committed:
refactor(terraform): Use path.module for robust source path

Updates the archive_file data source to use the `${path.module}` variable.

a-custodio committed:
style(terraform): Format configuration files with terraform fmt

a-custodio committed:
style(terraform): Add final newline to function.tf

a-custodio committed:
refactor(terraform): Move provider configuration to providers.tf

a-custodio committed:
refactor(terraform): Move backend configuration to backend.tf

a-custodio committed:
refactor(terraform): Moves provider version constraints to versions.tf

a-custodio committed:
style(terraform): Add final newline to secret_manager.tf

a-custodio committed:
refactor(terraform): Use random_id for unique bucket names

Updates the Terraform configuration to generate GCS bucket names dynamically by appending a random suffix. This prevents naming conflicts and makes the infrastructure code more reusable.

a-custodio committed:
refactor(terraform): Use path.module for robust file paths

Updates the `google_storage_bucket_object` resource to use the `${path.module}` variable for the `source` attribute. This ensures the path to `config.yaml` is resolved correctly

a-custodio committed:
style(terraform): Add final newline to variables.tf

a-custodio committed:
refactor(terraform): Automate service account creation and org permissions

This commit significantly improves the automation of the Terraform script:

- Creates the service account (`google_service_account`) directly within Terraform instead of requiring a pre-existing one.
- Automates the assignment of the `roles/viewer` permission at the organization level (`google_organization_iam_member`).
- Removes the `service_account_email` variable and adds the `organization_id` variable to support these changes.

a-custodio committed:
docs: Update README for new Terraform workflow

Revises the documentation to replace manual gcloud instructions with the new Terraform deployment process. Updates prerequisites, IAM roles, and local development sections to match the current architecture.

guilherme-hosoda-cit commented on infra/function.tf:
```diff
@@ -0,0 +1,71 @@
+  source_dir  = "${path.module}/../"
+  output_path = "/tmp/zombie-watcher-source.zip"
+  excludes = [
+    "terraform/",
```
Change excludes pattern to match `infra` folder instead of `terraform`.

```suggestion
    "infra/",
```

a-custodio replied:
Fixed!

guilherme-hosoda-cit commented on .gitignore:
```diff
@@ -572,3 +572,9 @@ result.json
+*.tfvars
+
+# Terraform lock file
+.terraform.lock.hcl
```
Check with the team if we should commit the `.terraform.lock.hcl` file.

a-custodio committed:
Changing excludes pattern to match infra folder instead of terraform

guilherme-hosoda-cit reviewed (approved):
LGTM
